### PR TITLE
Extract common Book object initialisation logic

### PR DESCRIPTION
**Pull request by ephphatha**
This really is a work in progress, trying to document small chunks to the best of my understanding so this can be split into specialised object classes or components in the future.

### CODE REVIEW

File: Source/objects.h [hunk 1]
```diff
@@ -58,28 +63,64 @@ struct ObjectStruct {
...
 	/**
-	 * @brief Convenience function for SetMapRange(Point, Point, int)
-	 * @param mapRange A rectangle defining the top left corner and size of the affected region
-	 * @param v Object subtype/discriminator
+	 * @brief Marks an object which was spawned from a sublevel in response to a lever activation.
+	 * @param mapRange The region which was updated to spawn this object.
+	 * @param leverID The id (*not* an object ID/index) of the lever responsible for the map change.
 	 */
-	constexpr void SetMapRange(Rectangle mapRange, int v)
+	constexpr void InitializeLoadedObject(Rectangle mapRange, int leverID)
```
Comment (ephphatha) [inline Source/objects.h:118]: This logic is also used for the quest objects in setmaps.cpp, I went with this name and description as the primary use seemed to be in objects.cpp:LoadMapObjects()

File: Source/objects.h [hunk 2]
```diff
@@ -58,28 +63,64 @@ struct ObjectStruct {
...
+	 *
+	 * Books of this type use a generic message (see OperateSChambBook()) compared to the more specific quest books
+	 * initialized by IntializeQuestBook().
+	 *
+	 * @param mapRange The region to be updated when this object is activated.
+	*/
+	constexpr void InitializeBook(Rectangle mapRange)
+	{
+		SetMapRange(mapRange);
+		_oVar6 = _oAnimFrame + 1; // Save the frame number for the open book frame
```
Comment (ephphatha) [inline Source/objects.h:97]: I'm pretty sure using `_oVar6` this way is unnecessary, it's used in OperateBookLever and OperateSChambBk to determine if the player has interacted with the object, but this check can also be accomplished by looking at the quest state (done earlier in OperateBookLever) or using this as a bool value. Changing that would require adding logic when saving/loading for compatibility.

File: Source/setmaps.cpp [hunk 3]
```diff
@@ -94,25 +94,30 @@ ObjectStruct &ObjectAtPosition(Point position)
...
-	ObjectAtPosition({ 46, 35 }).SetMapRange({ 8, 1 }, { 15, 11 }, 3);
-	ObjectAtPosition({ 49, 53 }).SetMapRange({ 8, 1 }, { 15, 11 }, 3);
-	ObjectAtPosition({ 27, 53 }).SetMapRange({ 8, 1 }, { 15, 11 }, 3);
+	constexpr Rectangle region1 { { 20, 7 }, { 3, 3 } };
+	ObjectAtPosition({ 64, 34 }).InitializeLoadedObject(region1, 1);
+
+	constexpr Rectangle region2 { { 20, 14 }, { 1, 2 } };
+	ObjectAtPosition({ 64, 59 }).InitializeLoadedObject(region2, 2);
+
+	constexpr Rectangle region3 { { 8, 1 }, { 7, 10 } };
```
Comment (AJenbo) [inline Source/setmaps.cpp:103]: would be nice if we could name them `smallSecretRoom`, `gate` and `largeSecretRoom`

**Review by AJenbo — APPROVED** [review]

File: Source/objects.h (hunk 2 above)
Comment (AJenbo) [reply Source/objects.h:97]: I do prefer isolating the bad architecture in the save/load routine

File: Source/setmaps.cpp (hunk 3 above)
Comment (ephphatha) [reply Source/setmaps.cpp:103]: Thanks heaps, I hadn't got around to working out what these actually were. I'll update the PR now.

**Review by AJenbo — APPROVED** [review]

**Pull request merged**